File: Java/src/examenes/exMarzo2019JavierFriasSerrano/ejercicio2/Rectangulo.java
package examenes.exMarzo2019JavierFriasSerrano.ejercicio2;

/**
 * Clase que define la creacion de Rectangulos
 * 
 * <p>
 * Enunciado Examen Módulo Programación 5 de Abril de 2019
 * </p>
 * <p>
 * 2.- (Java) Crea la clase Rectángulo de forma que:
 * </p>
 * <ul>
 * <li>• Un objeto de esta clase se construye pasándole el ancho y el alto.
 * Ninguno de los dos atributos puede ser menor o igual a cero ni mayor que
 * diez, en esos casos se debe lanzar la excepción ArithmeticException.</li>
 * <li>• Mediante getters y setters permite que se acceda y se modifique el
 * ancho y el alto del rectángulo teniendo en cuenta las restricciones en cuanto
 * a las dimensiones del apartado anterior.</li>
 * <li>• Al imprimir en pantalla un objeto de la clase usando System.out.print
 * se debe dibujar el rectángulo por la pantalla (de manera similar a como se
 * imprime un cuadrado en la página 130 del libro Aprende Java con
 * Ejercicios).</li>
 * <li>• Crea la clase Cuadrado como subclase de Rectángulo. Le debes añadir a
 * su comportamiento la posibilidad de comparar objetos cuadrados entre sí.</li>
 * <li>• Crea los programas de test correspondientes a ambas clases. Debes
 * provocar que se lance la excepción y capturarla.</li>
 * </ul>
 * 
 * @author Francisco Javier Frías Serrano
 * @version 1.0
 */
public class Rectangulo {

  private int ancho;
  private int alto;

  /**
   * Constructor
   * 
   * @param ancho
   * @param alto
   */
  public Rectangulo(int ancho, int alto) {
    setAncho(ancho);
    setAlto(alto);
  }

  /**
   * Constructor por defecto. Crea un rectángulo de 5x8
   */
  public Rectangulo() {
    setAncho(1);
    setAlto(1);
  }

  /**
   * Devuelve el ancho
   * 
   * @return el ancho
   */
  public int getAncho() {
    return ancho;
  }

  /**
   * Devuelve el alto
   * 
   * @return el alto
   */
  public int getAlto() {
    return alto;
  }

  /**
   * Establece el ancho
   * 
   * @param ancho
   *          a ancho establecer
   */
  public void setAncho(int ancho) {
    if (ancho <= 0 || ancho > 10) {
      throw new ArithmeticException("Ancho inválido");
    }
    this.ancho = ancho;
  }

  /**
   * Establece el alto
   * 
   * @param alto
   *          el alto a establecer
   */
  public void setAlto(int alto) {
    if (alto <= 0 || alto > 10) {
      throw new ArithmeticException("Alto inválido");
    }
    this.alto = alto;
  }

  /**
   * Pinta el rectangulo
   */
  @Override
  public String toString() {
    int i, espacios;
    String resultado = "";

    for (i = 0; i < this.ancho; i++) {
      resultado += "██";
    }
    resultado += "\n";

    for (i = 1; i < this.alto - 1; i++) {
      resultado += "██";
      for (espacios = 1; espacios < this.ancho - 1; espacios++) {
        resultado += "  ";
      }
      resultado += "██\n";
    }

    for (i = 0; i < this.ancho; i++) {
      resultado += "██";
    }
    resultado += "\n";

    return resultado;
  }

}

File: Python/exMarzo2019JavierFriasSerranoPython/ejercicio3/Rectangulo.py
'''
3. Haz el ejercicio anterior en Python. Enunciado del ejercicio anterior:

• Crea la clase Rectángulo de forma que:
• Un objeto de esta clase se construye pasándole el ancho y el alto. Ninguno de los dos atributos
puede ser menor o igual a cero ni mayor que diez, en esos casos se debe lanzar la excepción
ArithmeticException.
• Mediante getters y setters permite que se acceda y se modifique el ancho y el alto del
rectángulo teniendo en cuenta las restricciones en cuanto a las dimensiones del apartado
anterior.
• Al imprimir en pantalla un objeto de la clase usando System.out.print se debe dibujar el
rectángulo por la pantalla (de manera similar a como se imprime un cuadrado en la página 130
del libro Aprende Java con Ejercicios).
• Crea la clase Cuadrado como subclase de Rectángulo. Le debes añadir a su comportamiento
la posibilidad de comparar objetos cuadrados entre sí.
• Crea los programas de test correspondientes a ambas clases. Debes provocar que se lance la
excepción y capturarla.

__author__: Francisco Javier Frías Serrano
'''

class Rectangulo:

    def __init__(self, ancho, alto):
        """
        Constructor de la clase rectangulo, recibe por parametros dos enteros que conforman
        el ancho y el alto. Se verifican que los parametros sean correctos para poder asignarlos.
        """
        self.__ancho = ancho
        self.__alto = alto

    @property
    def ancho(self):
        return self.__ancho

    @ancho.setter
    def ancho(self, an):
        Rectangulo.verifica_lado(an)
        self.__ancho = an

    @property
    def alto(self):
        return self.__alto

    @alto.setter
    def alto(self, al):
        Rectangulo.verifica_lado(al)
        self.__alto = al

    def __str__(self):
        """
        Devuelve una cadena de texto con el rectangulo pintado, de manera
        que cuando se muestre el objeto en pantalla, se vea su representacion
        grafica.
        """
        cadena = ""
        cadena += "*"*self.ancho + "\n"
        if(self.alto == 2):
            cadena += "*"*self.ancho + "\n"
        elif(self.alto > 2):
            for i in range(0, self.alto - 2):
                cadena += "*"
                cadena += " "*(self.ancho - 2)
                cadena += "*\n"
            cadena += "*"*self.ancho + "\n"
        return cadena

    @staticmethod
    def verifica_lado(num):
        """
        Metodo estatico de la clase cuadrado, que comprueba que el lado se encuentre
        entre 1 y 10 y que sea un entero. Si no es un entero lanzara un TypeError y si
        no comprende los valores entre 1 y 10, lanzara un ArithmeticError.
        Nos servira para verificar tanto el alto, como el ancho del Rectangulo.
        """
        if not isinstance(num, int):  # lado no entero
            raise TypeError("Lado no entero", num)  # Lanzo esta excepcion si el parametro introducido no es un entero.
        if (num <= 0 or num > 10):
            raise ArithmeticError()  # Lanzo esta excepcion que es similar al ArithmeticException de Java


if __name__ == '__main__':
    try:
        rectangulo1 = Rectangulo(3, 4)
        rectangulo2 = Rectangulo(5, 10)
        
        # rectangulo3 = Rectangulo(11,3) # Descomentar para comprobar la excepcion la linea de abajo capturamos el ArithmeticError
        # rectangulo4 = Rectangulo(4.5,8) # Descomentando la linea de abajo capturamos el TypeError
        
        print("Pintamos el rectangulo 1 cuyos datos son:")
        print("Ancho: " + str(rectangulo1.ancho))
        print("Alto: " + str(rectangulo1.alto))
        print(rectangulo1)
        
        print("\n--------------------------------------\n")
        
        print("Pintamos el rectangulo 2 cuyos datos son:")
        print("Ancho: " + str(rectangulo2.ancho))
        print("Alto: " + str(rectangulo2.alto))
        print(rectangulo2)
        
        print("\n--------------------------------------\n")
        
        print("Modificamos el alto del rectangulo 2 a 4")
        rectangulo2.alto = 4
        
        print("Muestro 2 modificado:")
        print(rectangulo2)
        
    except ArithmeticError:
        print("Error. Los valores del rectangulo deben estar comprendidos entre 1 y 10")
    except TypeError:
        print("Error. El ancho y el alto deben ser valores enteros.")
    
